### PR TITLE
8268276: Base64 Decoding optimization for x86 using AVX-512

### PR DESCRIPTION
**Pull request by asgibbons**
This is a backport of JDK-8268276: Base64 Decoding optimization for x86 using AVX-512.

We have had numerous customer requests for this functionality to be backported due to the ~19x performance improvement.
 
Original patch does not apply cleanly to 17u, because of a bug that was identified in the original commit, but not fixed until the later Base64 Encode acceleration enhancement was complete.  The fix is on line 6039 of src/hotspot/cpu/x86/stubGenerator_x86_64.cpp and has been included in this PR.  All four "translatedX" registers needed to be checked for illegal characters in the input stream.

Risk: I view the risk of this backport to be minimal.  This code has been in use for many months with no bugs reported.
 
Testing: x86_64 build, affected tests, tier1, tier2
 
Thanks,
--Scott

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8268276](https://bugs.openjdk.org/browse/JDK-8268276): Base64 Decoding optimization for x86 using AVX-512

### Reviewers
 * [Paul Hohensee](https://openjdk.org/census#phh) (@phohensee - **Reviewer**)

### Contributors
 * Derek White `<drwhite@openjdk.org>`
 * Sandhya Viswanathan `<sviswanathan@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk17u-dev pull/840/head:pull/840` \
`$ git checkout pull/840`

Update a local copy of the PR: \
`$ git checkout pull/840` \
`$ git pull https://git.openjdk.org/jdk17u-dev pull/840/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 840`

View PR using the GUI difftool: \
`$ git pr show -t 840`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk17u-dev/pull/840.diff">https://git.openjdk.org/jdk17u-dev/pull/840.diff</a>

</details>

### CODE REVIEW

**Comment by dwhite-intel** [comment]
Hi Scott,

Can you post the diff between this backport and the original JDK-8268276 patch? I'm guessing it's pretty small, but it will be good to see.

Thanks!
 - Derek

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp [hunk 1]
```diff
@@ -5683,6 +5683,650 @@ address generate_avx_ghash_processBlocks() {
...
+      __ evpermt2b(translated0, input0, lookup_hi, Assembler::AVX_512bit);
+      __ evpermt2b(translated1, input1, lookup_hi, Assembler::AVX_512bit);
+      __ evpermt2b(translated2, input2, lookup_hi, Assembler::AVX_512bit);
+      __ evpermt2b(translated3, input3, lookup_hi, Assembler::AVX_512bit);
+
+      // OR all of the translations together to check for errors (high-order bit of byte set)
+      __ vpternlogd(input0, 0xfe, input1, input2, Assembler::AVX_512bit);
+
+      __ vpternlogd(input3, 0xfe, translated0, translated1, Assembler::AVX_512bit);
+      __ vpternlogd(input0, 0xfe, translated2, translated3, Assembler::AVX_512bit);
```
Comment (asgibbons) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:6039]: This line was changed from

      __ vpternlogd(input0, 0xfe, translated1, translated2, Assembler::AVX_512bit);

It was changed to:

      __ vpternlogd(input0, 0xfe, translated2, translated3, Assembler::AVX_512bit);

**Comment by asgibbons** [comment]
@dwhite-intel Did I answer your question to your satisfaction?  I'd appreciate a review if you have time.  Thanks!

**Comment by asgibbons** [comment]
So, the original patch applied cleanly, then I included the two bugfixes that were subsequently found (JDK-8273108 and JDK-8269404), one relating to improper illegal character discovery and the other relating to writing past end-of-string.

**Comment by dwhite-intel** [comment]
This looks good to me. Still need a (R)eviewer.

**Review by phohensee — CHANGES_REQUESTED** [review]
Looks like you didn't backport the change to .gitignore. Seems unrelated to the rest of the patch, but harmless enough.

Other than that, lgtm.

**Comment by asgibbons** [comment]
Thanks, @phohensee .  This is an unneeded artifact of me using VS for editing.  I do not believe it affects anything.

**Review by phohensee — APPROVED** [review]
Yes, it doesn't affect anything, but might be a merge conflict down the line. Let's see what the approvers say. Lgtm.

**Comment by asgibbons** [comment]
Trying to do dependent PRs instead of piecemealing.  Let me know if I'm doing something wrong.

**Pull request closed without merging**